petastorm/unischema.py

```diff
@@ -269,24 +287,69 @@ def as_spark_schema(self):
 
         return sql_types.StructType(schema_entries)
 
-    def make_namedtuple(self, **kargs):
-        """Returns schema as a namedtuple type intialized with arguments passed to this method.
-
-        Example:
-
-        >>> some_schema.make_namedtuple(field1=10, field2='abc')
+    def _get_field_names(self, results):
+        """ Figure out the field names from the returned results. Seems like can receive a namedtuple or a dictionary.
+        The dictionary may be nested, in which case need to check unique fields from each nested element - i.e. if
+        two nested elements contain the same field name, only include it in the final list once.
         """
-        # TODO(yevgeni): verify types
-        typed_dict = dict()
-        for key in kargs.keys():
-            if kargs[key] is not None:
-                typed_dict[key] = kargs[key]
-            else:
-                typed_dict[key] = None
-        return self._get_namedtuple()(**typed_dict)
+        if "_fields" in dir(results):
+            field_names = results._fields
+        else:
+            field_names = results.keys()
+            if isinstance(field_names, KeysView):
+                # If received a dictionary, rather than a list of keys, grab keys from each item in the dictionary,
+                # then remove any duplicates.
+                field_names = []
+                for key in results.keys():
+                    field_names = field_names + list(results[key]._fields)
+                # Remove duplicates.
+                field_names = list(set(field_names))
+        return field_names
+
+    def verify_fields(self, results, expected_fields=None):
+        """ Confirm that actual data matches layout that we expected. If not, warn user about differences.
+        I.e. if data doesn't contain fields that we expect, or does have fields that didn't expect, emit warning.
+        """
+        if "_is_verified" not in self.__dict__:
+            # Can happen if read older version from a file.
+            self._is_verified = False
+
+        if not self._is_verified:
+            # Can only verify if have some data.
+            if results:
+                actual_field_names = self._get_field_names(results)
+                missing_fields = []
```

Comment:
Working with sets could be clearer and possibly faster, as we would avoid python loops and n^2 complexity. Something like this?
```python
actual = set(self._get_field_names(results))
expected  = set(self._fields.keys())
missing = expected - actual
unexpected= actual - expected
```